Pause Warp Enhancement

mckinlee opened the pull request:
Original PR: #3212

This enhancement allows players to initiate warp songs directly from the pause menu, specifically from the quest status subscreen.

Changes:
~~Added PauseWarpState struct to manage the state of the PauseWarp feature.~~
~~Introduced IsPauseWarpEnabled() function to check if the PauseWarp feature is enabled.~~
~~Added InitiateWarp() function to initiate the warp process.~~
~~Added HandleWarpConfirmation() function to manage the warp song confirmation.~~
~~Implemented PauseWarp_Main() function to handle the initiation of the warp process from the pause menu.~~
~~Implemented PauseWarp_Idle() function to manage the state and handle edge cases.~~
~~Added cooldowns and state transitions to ensure proper functionality and user experience.~~

Edit:

I have rewritten this enhancement from the ground up. Here are the changes:

1. Cleaned up the code to make it easier to read and mess with.
2. Added lists for song messages, fanfares, and entrance locations, so it's clear what's connected to what.
3. Made starting a warp simpler with ActivateWarp, setting up the song and effects in one go.
4. PauseWarp_Execute now checks if we're ready to warp and handles it differently if you're playing with randomizer settings.
5. Picking a song to warp with from the pause menu is now straightforward with PauseWarp_HandleSelection.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237503510.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237507347.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237507354.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237508071.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237508325.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237511267.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1237519740.zip)
<!--- section:artifacts:end -->

mckinlee committed:
Pause Warp Enhancement

This commit introduces the PauseWarp mod, a feature that allows players to warp to different locations in the game directly from the pause menu.

- Add PauseWarpState structure to manage flags and cooldowns for the pause warp feature.
- Implement IsStateValid function for state validation.
- Implement ResetStateFlags function to reset all state flags to default values.
- Add InitiateWarp function to handle the initiation of warp sequences.
- Implement HandleWarpConfirmation function to confirm and execute warp actions.
- Implement HandleCooldowns function to manage various cooldown timers.
- Add PauseWarp_Main function as the main logic, called every frame to handle pause warp functionality.
- Map warp song messages to in-game text messages.

mckinlee committed:
Warp Song Check

-Now if you do not have a warp song you won't be able to select the empty slot and still teleport.

mckinlee committed:
Added Audio Fanfares and Changed stateFlag1 to PLAYER_STATE1_IN_CUTSCENE

-When selecting a warp song the audio for the applicable warp song will now play for a extra vanilla feel.
-Changed the stateFlag1 because previously it just disabled input allowing enemies to harm you. Now that won't happen because the game is put into a cutscene state.

mckinlee committed:
Feedback Update

-A new hook was created 'OnPauseMenu' so now PauseWarp_Main is only called when the pause menu is open
-Moved pauswarp.c to the Enhancements folder
-Removed from graph.c

PR Change:
Changing to the main branch instead of sulu

mckinlee committed:
Feedback Update #2

-Introduced new function 'PauseWarp_Idle' now that 'PauseWarp_Main' is no longer called every frame
-Added C wrapper to access 'GameInteractor::IsSaveLoaded' and scrapped the 'IsStateValid' function
-Added 'PauseWarp_Idle' to the the 'RegisterPauseWarp' function
-Refactored the code some

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-pausewarp

mckinlee committed:
Linux Compile Issue

-Added a missing header that was causing a compile issue for linux
-Hopefully, it won't crash

mckinlee committed:
Minor Bug Fix

-Now link won't get soft locked when warping to the same location twice

CasperMcFadden95 commented:
Awesome work!

Can you also add this functionality to the Map? There's a decide button which is disabled but that could be used to warp somwhere.

![ootmap](https://cdn.mobygames.com/screenshots/15709624-the-legend-of-zelda-ocarina-of-time-nintendo-64-hyrule-map.png)

Thanks

PurpleHato commented:
> Awesome work!
> 
> Can you also add this functionality to the Map? There's a decide button which is disabled but that could be used to warp somwhere.
> 
> ![ootmap](https://camo.githubusercontent.com/9a773bff5cb7ba9b5577b05d4ead84521c36c9ccdce3e27999b3219f1dc84cb1/68747470733a2f2f63646e2e6d6f627967616d65732e636f6d2f73637265656e73686f74732f31353730393632342d7468652d6c6567656e642d6f662d7a656c64612d6f636172696e612d6f662d74696d652d6e696e74656e646f2d36342d687972756c652d6d61702e706e67)
> 
> Thanks

The point of this enhancement is to have warp songs (where warp pedestals are) being easy usable;
The map contains points that doesn't have pedestal, hence why it's not a good idea to do that an d for multiple aspects
- Zone names become incoherant (since only 2 have the name of the zones used by warp songs, the rests are literally unrelated to warps and break logic
- Do you warp then on the zone name like Lon Lon Ranch directly? If yes, you need to add that option to use that logic in randomizer to not break it ?
- Do you create custom points, with custom texture when it's enabled and basically redo a complete new map ?

That's why it's not viable to do as a QoL but rather a code mod made by the community in my opinion

Malkierian commented:
I don't like that idea for several reasons. First, there are more locations there than warp songs, and differentiating between locations you can and can't warp to would be awkward. Second, warp shuffle would make that even more difficult because then you'd have to know where the warp song was taking you and then apply that to the location on the map. At the very least, it feels like a different PR to me, if not mod territory, depending on how you wanted to apply it.

CasperMcFadden95 commented:
Thanks I understand. Can you also make this mod play the non-warp songs? I find it annoying that I have to press the notes. Thanks

mckinlee commented:
> Thanks I understand. Can you also make this mod play the non-warp songs? I find it annoying that I have to press the notes. Thanks

Thanks you for your comments! I appreciate it.

That might be something I would consider working on in the future.. maybe, but it would have to be a totally different PR and probably something more akin to the Skip Scarecrow Song enhancement. The groundwork is set with this PR, so it shouldn't be to bad for someone to come along and do that if they wished.  I think I've done all I want to do with this current feature and have moved on to another project. I'm trying to implement a banking system similar to the one in MM.

mckinlee committed:
Merge branch 'HarbourMasters:develop' into develop-pausewarp

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into develop-pausewarp

mckinlee committed:
Update libultraship

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into develop-pausewarp

mckinlee committed:
Revert "Update libultraship"

This reverts commit 746fc234795c06261a4fb69484f4656676f1eaaa.

mckinlee commented:
After merging to latest dev build, game now soft locks unless the player has a Ocarina in their inventory. Will work on a fix when I can.

mckinlee committed:
Bug Fix
-Added more checks to ensure vanilla behavior when a Ocarina is not in the players inventory.

mckinlee commented:
Bug fixed. This is merge ready again.

mckinlee committed:
Merge branch 'develop' into develop-pausewarp

mckinlee commented:
Updated to MacReady Bravo (8.0.1)

mckinlee committed:
Merge branch 'develop' into develop-pausewarp

mckinlee committed:
WIP

mckinlee committed:
Done unless I'm missing headers

briaguya0 commented on soh/soh/Enhancements/game-interactor/GameInteractor.cpp:
```diff
@@ -45,6 +45,11 @@ bool GameInteractor::IsSaveLoaded() {
     return (gPlayState == NULL || player == NULL || gSaveContext.fileNum < 0 || gSaveContext.fileNum > 2) ? false : true;
 }
 
+extern "C" bool IsSaveLoadedWrapper() {
+    GameInteractor* gameInteractor = GameInteractor::Instance;
+    return gameInteractor->IsSaveLoaded();
+}
```
i think following the naming patterns used for other extern c GI things might make sense here

https://github.com/HarbourMasters/Shipwright/blob/8a14fea94c07332bcf61a135076d930073e1d2da/soh/soh/Enhancements/game-interactor/GameInteractor_Hooks.h#L3-L8

https://github.com/HarbourMasters/Shipwright/blob/8a14fea94c07332bcf61a135076d930073e1d2da/soh/soh/Enhancements/game-interactor/GameInteractor_Hooks.cpp#L9-L11

so something like

```suggestion
extern "C" bool GameInteractor_IsSaveLoaded() {
    return GameInteractor::Instance->IsSaveLoaded();
}
```

briaguya0 reviewed (approved):
commented about a name change but i don't see that as blocking

mckinlee committed:
now we done

mckinlee committed:
Merge remote-tracking branch 'origin/pausewarpredo' into develop-pausewarp

mckinlee committed:
clean up, these arn't needed anymore

mckinlee commented:
> commented about a name change but i don't see that as blocking

Again, thanks for reviewing! Like I mentioned in discord, this was a utter mess due to my inexperience with coding at the beginning.. Given that, I have committed a rewrite that is a lot cleaner, works with rando now, and removes that IsSaveLoadedWrapper. It was much better to just use GameInteractor::IsSaveLoaded directly in mods.cpp. This should be re-reviewed whenever yall get a chance.

garrettjoecox commented on soh/soh/Enhancements/game-interactor/GameInteractor.h:
```diff
@@ -221,6 +221,7 @@ class GameInteractor {
 
     DEFINE_HOOK(OnFileDropped, void(std::string filePath));
     DEFINE_HOOK(OnAssetAltChange, void());
+    DEFINE_HOOK(OnPauseMenu, void());
```
I think this maybe makes more sense as `OnKaleidoUpdate`

I also assume that this means `OnGameFrameUpdate` does not run with kaleido open? So you couldn't just also use `OnGameFrameUpdate` and instead just check some sort of pauseCtx value?

mckinlee replied:
Nah, OnGameFrameUpdate does run with Kaleido open. You're right, I could of just checked if pauseCtx->pageIndex == PAUSE_QUEST. However, on my original PR for this @jbodner09  liked the idea of having a hook specific to the pause menu for custom item mods. That comment is [here](https://github.com/HarbourMasters/Shipwright/pull/3212#issuecomment-1724958636). That's why I kept it in, but it's up to yall. It's not necessary for my use case so I could remove it.

garrettjoecox replied:
On second thought, `OnGameFrameUpdate` is not something we want to use often, so this may be better after all

mckinlee replied:
I agree. I tried to figure out a way around using it for PauseWarp_Execute, but couldn't figure out a easy way to do it. I've made the checks pretty specific so it shouldn't cause any issue anyway.

garrettjoecox replied:
Yea should be fine, I still stand by my original rename suggestion though 👍

mckinlee replied:
Fine by me and I just changed it in the latest commit.

garrettjoecox reviewed (approved):
This is borderline mod territory, but the fact it's well isolated and the only touch point in source is a generic hook in my mind makes it fine.

One comment/question on the new hook

mckinlee committed:
Rename OnPauseMenu to OnKaleidoUpdate
